Add EFS storage support to ECS Clusters

roleyfoley opened the pull request:
Allows for an EFS volume to be created and attached to each node in an ECS cluster. Data is then persisted and shared in the cluster.

ml019 commented on aws/templates/solution/solution_ecs.ftl:
```diff
@@ -53,6 +53,47 @@
+            mode=listMode
+            tier=tier
+            component=component 
+            extensions="efs"
```
not sure we need "efs" extension - more for when we use getOccurrences

roleyfoley replied:
Since the security group is created as part of the ECS creation the component details are for the ECS instance instead of EFS. Since this security group is for access to the EFS instance created as part of the ECS component I added it, otherwise the group name would have matched the ecs cluster not the efs instance

ml019 replied:
ok - makes sense

ml019 commented on aws/templates/solution/solution_ecs.ftl:
```diff
@@ -53,6 +53,47 @@
+            ecs.ClusterWideStorage?has_content && ecs.ClusterWideStorage ]
+
+        [#assign ecsEFSVolumeId = formatEFSId(tier, component)]
+        [#assign ecsEFSVolumeName = formatName( tier, component, "efs")]
```
Doesn't need efs in the name I think - it will be shown in the console under EFS - but it should be the fulkl component name including product/segment

roleyfoley replied:
Name updated to formatFullComponentName and "efs" removed

ml019 commented on aws/templates/solution/solution_ecs.ftl:
```diff
@@ -53,6 +53,47 @@
+            extensions="efs"
+            /]
+        
+        [@createSecurityGroupIngress
```
DO we need this? Maybe ingress rules added when the efs is mounted - so by defualt the efs itself has no ingress rules

roleyfoley replied:
Two security group components are added, one which creates a security group that controls the EFS eni interfaces and the Ingress update is for the ECS cluster security group. 

Since the two are created as part of the same component then they are added together.

ml019 replied:
ok - my bad

ml019 commented on aws/templates/solution/solution_ecs.ftl:
```diff
@@ -53,6 +53,47 @@
+            tier=tier
+            efsId=ecsEFSVolumeId
+            securityGroups=ecsEFSSecurityGroupId
+            dependencies=[ecsEFSVolumeId,ecsEFSSecurityGroupId]
```
do we explicitly need dependencies given both are parameters to macro anyway so CF will likely figure them out for itself...

roleyfoley replied:
Removed

ml019 commented on aws/templates/solution/solution_ecs.ftl:
```diff
@@ -198,14 +240,25 @@
                                     }) +
+                                attributeIfTrue(
+                                    "02ConfigureEFSClusterWide",
+                                    ecs.ClusterWideStorage?has_content && ecs.ClusterWideStorage == true,
```
don't check has_content - it always will

roleyfoley replied:
Fixed

ml019 commented on aws/templates/solution/solution_efs.ftl:
```diff
@@ -0,0 +1,44 @@
+[#-- EFS --]
```
Review in light of comments above

roleyfoley commented on aws/templates/resource/resource_efs.ftl:
```diff
@@ -0,0 +1,74 @@
+        ARN_ATTRIBUTE_TYPE : { 
+            "UseRef" : true
+        },
+        NAME_ATTRIBUTE_TYPE : { 
```
Removed Name attribute as it doesn't actually have one

roleyfoley commented on aws/templates/resource/resource_efs.ftl:
```diff
@@ -0,0 +1,74 @@
+
+        [#list getSubnets(tier) as subnet ]
+
+            [#assign efsMountTargetId = formatDependentEFSMountTargetId(
```
Fixed

ml019 replied:
Doesn't look like zone is initialised at this point? 

Either we copy logic from getSubnets here (as it iterates through the zones) or we enhance getSubnets to support optionally passing the required zone(s) in as a third parameter?

roleyfoley commented on aws/templates/solution/solution_ecs.ftl:
```diff
@@ -53,6 +53,47 @@
     [/#if]
 
+    [#if deploymentSubsetRequired("efs", true) && 
+            ecs.ClusterWideStorage?has_content && ecs.ClusterWideStorage ]
```
Fixed

ml019 commented on aws/templates/resource/resource_efs.ftl:
```diff
@@ -0,0 +1,71 @@
+
+[#macro createEFSMountTarget mode tier efsId securityGroups dependencies="" ]
+
+    [#if getSubnets(tier)?has_content ]
```
if not necessary as getSubnets always returns a (possibly empty) array

roleyfoley commented:
Added an includeZone option to GetSubnets. 
This returns an object which lists the subnet Id ( either getReference or Id ) depending on asReferences 

Updated the EFS mount targets to use the new GetSubnets options

ml019 commented on aws/templates/solution/solution_ecs.ftl:
```diff
@@ -198,14 +238,25 @@
                                     }) +
+                                attributeIfTrue(
+                                    "02ConfigureEFSClusterWide",
+                                    ecs.ClusterWideStorage == true,
```
If it is a boolean so no need to compare to true
